lib/dockerregistry/checks.go

```diff
@@ -0,0 +1,159 @@
+type ImageRemovalCheck struct {
+	CurrentGitRepo url.URL
+	MasterSHA      plumbing.Hash
+	PullRequestSHA plumbing.Hash
+}
+
+// MKRealImageRemovalCheck returns an instance of ImageRemovalCheck.
+func MKRealImageRemovalCheck(
+	currentGitRepoStr string,
+) (*ImageRemovalCheck, error) {
+	currentGitRepo, err := url.Parse(currentGitRepoStr)
+	if err != nil {
+		return nil, fmt.Errorf("Could not parse the git repo url: %v", err)
+	}
+	masterSHAStr := os.Getenv("PULL_BASE_SHA")
+	if masterSHAStr == "" {
+		return nil, fmt.Errorf("The PULL_BASE_SHA environment variable " +
+			"doesn't exist")
+	}
+	pullRequestSHAStr := os.Getenv("PULL_PULL_SHA")
+	if pullRequestSHAStr == "" {
+		return nil, fmt.Errorf("The PULL_PULL_SHA environment variable " +
+			"doesn't exist")
+	}
+	masterSHA := plumbing.NewHash(masterSHAStr)
+	pullRequestSHA := plumbing.NewHash(pullRequestSHAStr)
+	return &ImageRemovalCheck{
+		*currentGitRepo,
+		masterSHA,
+		pullRequestSHA,
+	}, nil
+}
+
+// Run executes ImageRemovalCheck on a set of promotion edges.
+// Returns an error if the pull request removes images from the
+// promoter manifests.
+func (check *ImageRemovalCheck) Run(
+	edges map[PromotionEdge]interface{},
+) error {
+	var err error
+	// Get the Git repo the current working directory
+	r, err := gogit.PlainOpen(check.CurrentGitRepo.String())
+	if err != nil {
+		return fmt.Errorf("Could not open the Git repo: %v", err)
+	}
+
+	// Check out the master branch of the current Git repo to get the
+	// master branch's state of the promoter manifests
+	w, err := r.Worktree()
+	if err != nil {
+		return fmt.Errorf("Could not create Git worktree: %v", err)
+	}
+	err = w.Checkout(&gogit.CheckoutOptions{
+		Hash:  check.MasterSHA,
+		Force: true,
+	})
+	if err != nil {
+		return fmt.Errorf("Could not checkout the master branch of the Git"+
+			" repo: %v", err)
+	}
+
+	// Parse the manifests and generate the promotion edges
+	mfests, err := ParseThinManifestsFromDir(check.CurrentGitRepo.String())
+	if err != nil {
+		return fmt.Errorf("Could not parse manifests from the directory: %v",
+			err)
+	}
+	masterEdges, err := ToPromotionEdges(mfests)
+	if err != nil {
+		return fmt.Errorf("Could not generate promotion edges from promoter"+
+			" manifests: %v", err)
+	}
+
+	// Reset the working Git repo by checking out the pull request
+	err = w.Checkout(&gogit.CheckoutOptions{
+		Hash:  check.PullRequestSHA,
+		Force: true,
+	})
+	if err != nil {
+		return fmt.Errorf("Could not checkout the pull request branch of the"+
+			" Git repo %v: %v",
+			check.CurrentGitRepo, err)
+	}
+
+	return check.Compare(masterEdges, edges)
+}
+
+// Compare is a function of the ImageRemovalCheck that handles
+// the comparison of the pull requests's set of promotion edges and
+// the master branch's set of promotion edges.
+func (check *ImageRemovalCheck) Compare(
+	edgesMaster map[PromotionEdge]interface{},
+	edgesPullRequest map[PromotionEdge]interface{},
+) error {
+	type ImageInfo struct {
```

Comment:
This shouldn't be necessary... why can't we just compare edges directly?

In general you should be very conservative about introducing new data structures, even temporary ones. Imagine now testing for `ImageInfo` --- one disadvantage (among others) is you can't test it unless you move it out of this function.